### PR TITLE
Fix #782 -- use hyperlinks in report on all activity

### LINKED ISSUE #782: Report on all activity should use hyperlinks for workshops
On https://amy.software-carpentry.org/workshops/reports/all_activity_over_time, the references to workshops should be hyperlinks, not just plain-text slugs.

### PR DESCRIPTION
**Pull request by chrismedrela**
Fixes #782.

### CODE REVIEW

File: api/views.py [hunk 1]
```diff
@@ -266,16 +267,17 @@ def _only_latest_date(self, iterable):
...
-        format_ = (request.query_params.get('format') or format or '').lower()
+        if request is not None:
+            format_ = (request.query_params.get('format') or format or '').lower()
 
-        if format_ in self.formats_requiring_lists:
-            # list-ify the generator for renderers requiring lists
-            return list(iterable)
+            if format_ in self.formats_requiring_lists:
+                # list-ify the generator for renderers requiring lists
+                return list(iterable)
```
Comment (chrismedrela) [inline api/views.py]: @pbanaszkiewicz , could we always listify, that is, always return `list(iterable)`?

File: api/views.py [hunk 2]
```diff
@@ -430,16 +439,16 @@ def all_activity_over_time(self, request, format=None):
...
 
-        return Response({
+        return {
             'start': start,
             'end': end,
             'workshops': {
                 'SWC': swc_workshops.count(),
                 'DC': dc_workshops.count(),
-                'SWC,DC': swc_dc_workshops,
+                'SWC_or_DC': swc_dc_workshops,
```
Comment (chrismedrela) [inline api/views.py]: @pbanaszkiewicz, We change the API here. Is it use anywhere?

File: api/views.py (hunk 1 above)
Comment (pbanaszkiewicz) [inline api/views.py]: No, this is not a good solution performance-wise.

File: api/views.py [hunk 3]
```diff
@@ -268,16 +269,17 @@ def _only_latest_date(self, iterable):
...
             yield next_
 
-    def listify(self, iterable, request, format=None):
+    def listify(self, iterable, request=None, format=None):
         """Some renderers require lists instead of any iterables for rendering.
         This function conditionally turns iterables into lists based on the
         format requested by browser."""
         # choose either '?format=...' or '/url.format' or None
-        format_ = (request.query_params.get('format') or format or '').lower()
+        if request is not None:
```
Comment (pbanaszkiewicz) [inline api/views.py]: Why is this here?

File: api/views.py (hunk 2 above)
Comment (pbanaszkiewicz) [inline api/views.py]: It was used here: https://github.com/swcarpentry/amy/pull/1060/commits/6239fa34d6fb0c4798a5ac2831a5040623493ffe#diff-345991cbd83965baed2fa14d0e2f37ceL106

**Comment by pbanaszkiewicz** [comment]
:-1: because:
* missing explanation of changes
* actual changes make code harder to follow
* changes in data structure

File: api/views.py (hunk 3 above)
Comment (chrismedrela) [inline api/views.py]: You're right, this change is unnecessary.

Comment (pbanaszkiewicz) [inline api/views.py]: Sorry, I probably sounded rude, but I actually meant to ask what was this code doing and why, because you hadn't included any explanation.

**Comment by chrismedrela** [comment]
> changes in data structure

Do you mean changes in the schema of returned JSON?

> actual changes make code harder to follow

I completely agree, however I don't know how to make the code easier to follow. Any ideas?

**Comment by pbanaszkiewicz** [comment]
> Do you mean changes in the schema of returned JSON?

Yes.

> I completely agree, however I don't know how to make the code easier to follow. Any ideas?

I don't know what you're achieving with this code (== "I don't understand it"), so I don't know if we can do better.

File: api/views.py (hunk 3 above)
Comment (chrismedrela) [inline api/views.py]: But this change *really* was unnecessary, so I've reverted it. :-)

**Comment by chrismedrela** [comment]
> I don't know what you're achieving with this code (== "I don't understand it"), so I don't know if we can do better.

So far, what we did is we returned HTML page without data with a bit of JS that asks API for the data and, then, renders the data.

Now, everything is rendered on server. When triggering `all_activity_over_time` view, server asks API for the data. However, we don't make a request, we just call `ReportsViewSet.get_all_activity_over_time` to get the data as Python dictionary without network overhead. Then, we pass the dictionary to `all_activity_over_time.html` template where it's traversed. We had to change data structure, because django templates don't allow special characters when traversing the structure (i.e., you cannot write `{{ data.workshops['SWC,DC'] }}`).

Yes, I should have included a description like the above one in the commit message. Sometimes I forget that other people don't have access to my mind. :-)

Is there anything else you want me to clarify?

**Comment by pbanaszkiewicz** [comment]
> Is there anything else you want me to clarify?

Now it's clear to me, thank you. I'll get back to you about the solution soon.

**Comment by pbanaszkiewicz** [comment]
Okay, so here's I think the sanest solution I came up with so far.

1) Drop `listify` from `get_all_activity_over_time` and make it only return "pure" dict:

```python
    def get_all_activity_over_time(self, start, end):
        events_qs = Event.objects.filter(start__gte=start, start__lte=end)
        …

        # Workshops missing any of this data.
        # --- REMOVING COMMENT ABOUT JS FROM HERE ---
        missing_attendance = …

        return {
            'start': start,
            'end': end,
            'workshops': {
                'SWC': swc_workshops.count(),
                'DC': dc_workshops.count(),
                'SWC_or_DC': swc_dc_workshops,
                'WiSE': wise_workshops,
                'TTT': ttt_workshops,
                'self_organized': self_organized_workshops,
            },
            'instructors': {
                'SWC': {
                    'total': swc_total_instr,
                    'unique': swc_unique_instr,
                },
                'DC': {
                    'total': dc_total_instr,
                    'unique': dc_unique_instr,
                },
            },
            'learners': {
                'SWC': swc_total_learners,
                'DC': dc_total_learners,
            },
            'missing': {
                'attendance': missing_attendance,
                'instructors': missing_instructors,
            }
        }
```

2) use `listify` in `all_activity_over_time` view on `get_all_activity_over_time` output:

```python
    @list_route(methods=['GET'])
    def all_activity_over_time(self, request, format=None):
        """Workshops, instructors, and missing data in specific periods."""
        start, end = self._default_start_end_dates(
            start=request.query_params.get('start', None),
            end=request.query_params.get('end', None))

        data = self.get_all_activity_over_time(start, end)
        data['missing']['attendance'] = self.listify(…)
        data['missing']['instructors'] = self.listify(…)
        return Response(data, …?)
```

3) Instead of changing the original structure, extend it: (in `get_all_activity_over_time`)

```python
            'workshops': {
                'SWC': swc_workshops.count(),
                'DC': dc_workshops.count(),
                'SWC_or_DC': swc_dc_workshops,
                'SWC,DC': swc_dc_workshops,  # for backwards-compatibility
                'WiSE': wise_workshops,
                'TTT': ttt_workshops,
                'self_organized': self_organized_workshops,
                'self-organized': self_organized_workshops,  # for backwards-compatibility
            },
```

Now we should not have any problems with API structure, and we don't need any partials, and we keep your design (ie. server-side rendering).

**Comment by pbanaszkiewicz** [comment]
@chrismedrela is this ready?

**Comment by chrismedrela** [comment]
@pbanaszkiewicz yes.

File: api/views.py [hunk 4]
```diff
@@ -438,16 +442,18 @@ def all_activity_over_time(self, request, format=None):
...
             'end': end,
             'workshops': {
                 'SWC': swc_workshops.count(),
                 'DC': dc_workshops.count(),
                 'SWC,DC': swc_dc_workshops,
+                'SWC_or_DC': swc_dc_workshops,
                 'WiSE': wise_workshops,
                 'TTT': ttt_workshops,
                 'self-organized': self_organized_workshops,
+                'self_organized': self_organized_workshops,
```
Comment (pbanaszkiewicz) [inline api/views.py]: One small final bit: can you please add a comment here explaining why we need `SWC,DC` and `SWC_or_DC` (same for self organized)?

Comment (chrismedrela) [inline api/views.py]: @pbanaszkiewicz ping
